### PR TITLE
Add multicast tests and fix ConnectableObservable.

### LINKED ISSUE #657: comprehensive tests for multicast

### PR DESCRIPTION
**Pull request by staltz**
Resolves #657 and solves bug #678. Took some time because #678 was truly a nightmare.

### CODE REVIEW

**Comment by staltz** [comment]
By the way, @trxcllnt you might want to review this if I'm doing any performance mistake. At least there is one closure introduced there. I focused only on getting the bug fixed, we might want to do a second round to make it performant if there's opportunity for that.

**Comment by staltz** [comment]
I just rebased, but now as you can see, tests fail. And the fix isn't immediate. I'm solving some concurrency issues now in ConnectableObservable. I don't know exactly which change to master happened that affected this.

File: src/observables/ConnectableObservable.ts [hunk 1]
```diff
@@ -62,24 +92,60 @@ class RefCountObservable<T> extends Observable<T> {
 
   _subscribe(subscriber) {
     const connectable = this.connectable;
-    const subscription = connectable.subscribe(subscriber);
-    if (++this.refCount === 1) {
-      this.connection = connectable.connect();
+    const refCountSubscriber = new RefCountSubscriber(subscriber, this);
+    refCountSubscriber.myConnection = this.connection;
```
Comment (blesh) [inline src/observables/ConnectableObservable.ts]: why is this not added via the constructor above it?

File: src/observables/ConnectableObservable.ts [hunk 2]
```diff
@@ -62,24 +92,60 @@ class RefCountObservable<T> extends Observable<T> {
...
+
+  _complete() {
+    this._resetConnectable();
+    this.destination.complete();
+  }
+
+  _resetConnectable() {
+    const observable = this.refCountObservable;
+    const myConnection = this.myConnection;
+    if (myConnection && myConnection === observable.connection) {
```
Comment (blesh) [inline src/observables/ConnectableObservable.ts]: could probably just get the `connection` out of `observable` in a const above, `const connection = observable.connection`

File: src/observables/ConnectableObservable.ts (hunk 1 above)
Comment (staltz) [inline src/observables/ConnectableObservable.ts]: It can. What's more important is line 101.

File: src/observables/ConnectableObservable.ts (hunk 2 above)
Comment (staltz) [inline src/observables/ConnectableObservable.ts]: Will do.

**Comment by staltz** [comment]
@blesh I refactored it.

**Comment by blesh** [comment]
haha... staltz, this is perfect, but I have _one_ more request. Can you flatten the fix commits? :)

**Comment by staltz** [comment]
Yes sir I can. I made them separate in the first place because they hold a lot of story (just check the commit message) on how and why did I make those changes.

**Comment by staltz** [comment]
@blesh Flat it is.

**Comment by blesh** [comment]
:+1:  LGTM ... merging in 3... 2... 1...

**Comment by blesh** [comment]
merged with 20b9a8c... thanks @staltz!

**Comment by staltz** [comment]
That was one hell-ov-a PR.

**Pull request closed without merging**